src/subsystems/nodejs/builders/granular/default.nix

```diff
@@ -345,33 +465,39 @@
             exit 1
           fi
 
-          # configure typescript
-          if [ -f ./tsconfig.json ] \
-              && node -e 'require("typescript")' &>/dev/null; then
+          # configure typescript to resolve symlinks locally
+          # TODO is this really needed? Node doens't use it
```

Comment:
this option and its Node equivalent is only necessary if the tree of symlinks is incorrect. The way the node_modules are built now, each module sees only the modules it's supposed to see, and it should work that way in ts as well.

So if typescript needs the flag, it means it's accessing modules that it didn't request.

Not sure what the memory implications of this flag are, if it treats multiple symlinks to the same module as multiple copies of the module. That would also break module-scoped storage.

Reply:
OK, but I believe there are cases where upstream lock files do have incomplete dependency declarations. I think for example in package-lock v1, peerDependencies are not declared explicitly, instead it is assumed that the parent installs them. Therefore preserveSymlinks is required in order to find the parent. Otherwise a `../` will lead to the `/nix/store` path of the current module.

Reply:
But in v2 they are indeed declared. What this PR does is extract all deps each package needs and then resolves them the way node would resolve them. So peer deps are either resolved to the correct version, or missing from the project entirely.

Reply:
That's great. But we need to be downward compatible to the old package-lock.json versions as well. So we'd need to find a of solution for that.

Reply:
> I don't think there's a solution, the information is just missing and the logic we have in place for v1 doesn't work for e.g. apollo-server-errors, it can't find graphql.

As explained in https://github.com/nix-community/dream2nix/issues/198#issuecomment-1185377101 It worked just fine with the old build logic, assuming that `preserveSymlinks` is used currently by all tooling.
I just tried building `apollo-server-errors` as well with the default dream2nix flake and it builds just fine.
If I'm wrong please paste a flake, that demonstrates it.

I think what you are doing here is an improvement, that frees us from having to use `preserveSymlinks`, which is really great, because many build tools don't cope well with it.

But I don't think we should throw away the old installation logic, as long as the new logic only improves upon some use cases, but breaks others. Currently it breaks all lock file v1 projects, and requiring `resolveImpure` on all these projects is a significant trade-off.

I think we should either extend the build logic to also support v1, or we should keep the old build logic as an alternative.

Reply:
point taken, I can actually just do the v2 logic when the file is v2. I'll adjust.